### PR TITLE
A/B test package

### PR DESCRIPTION
**Pull request by clchinara**
Added A/B test client (#1 )

### CODE REVIEW

File: src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    private final Gson gson;
+
+    public ABTestClientImpl(KaleConfig kaleConfig) {
+        this.kaleConfig = kaleConfig;
+        this.httpClient = new OkHttpClient();
+        this.gson = new Gson();
+    }
+
+    @Override
+    public JsonObject getUniverseAllocation(String userId, String universeId) {
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java:29]: knp gk bikin object misal UniverseAllocation.  jd client yg pake nanti tinggal get property. biar lebih predictable

File: src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                userId, universeId);
+    
+        GetUniverseAllocationResponse response = cmd.execute();
+        JsonObject json = constructABTestJson(response);
+
+        return json;
+    }
+
+    @Override
+    public JsonArray getAllUniverseAllocations(String userId) {
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java:41]: idem

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GetAllUniverseAllocationsCommand extends KaleCommand<List<GetUniverseAllocationResponse>> {
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:19]: dibikin jg test nya sprti https://github.com/sayurbox/kale-client-java/blob/master/src/test/java/com/sayurbox/kale/featureflag/command/GetAllocateCommandTest.java

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.sayurbox.kale.exception.KaleException;
+import com.sayurbox.kale.abtest.client.GetUniverseAllocationResponse;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import java.io.UnsupportedEncodingException;
+
+public class GetUniverseAllocationCommand extends KaleCommand<GetUniverseAllocationResponse> {
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:16]: test nya

File: src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import com.sayurbox.kale.abtest.command.GetUniverseAllocationCommand;
+import okhttp3.OkHttpClient;
+
+public class ABTestClientImpl implements ABTestClient {
+
+    private final KaleConfig kaleConfig;
+    private final OkHttpClient httpClient;
+    private final Gson gson;
+
+    public ABTestClientImpl(KaleConfig kaleConfig) {
```
Comment (andrisasuke) [inline src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java:22]: dibikin testnya sprti https://github.com/sayurbox/kale-client-java/blob/master/src/test/java/com/sayurbox/kale/featureflag/FeatureFlagClientImplTest.java

File: src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java (hunk 1 above)
Comment (clchinara) [reply src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java]: Fixed here kak:
https://github.com/sayurbox/kale-client-java/pull/2/commits/1cbfa0d33101fa2cf27be6bf766380210f0aa249
https://github.com/sayurbox/kale-client-java/pull/2/commits/7e20bd2a6084e7fe32dc8efe9ae0ce78fd0f81ef

File: src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java (hunk 5 above)
Comment (clchinara) [reply src/main/java/com/sayurbox/kale/abtest/ABTestClientImpl.java:17]: Added here: https://github.com/sayurbox/kale-client-java/pull/2/commits/30fe04b591ffa6d8f3430fb4e76f3d5a363cf1ab

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java (hunk 4 above)
Comment (clchinara) [reply src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:14]: Added here: https://github.com/sayurbox/kale-client-java/pull/2/commits/2a2947762c002bf8711ee684bc65a105156b8129

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java (hunk 3 above)
Comment (clchinara) [reply src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:18]: Added here: https://github.com/sayurbox/kale-client-java/pull/2/commits/2a2947762c002bf8711ee684bc65a105156b8129

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public List<GetUniverseAllocationResponse> getFallback() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    protected Request createRequest() {
+        String url = String.format("%s/v1/abtest/allocation/%s",
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:38]: dibikin constant aja dibagian atas class ini @clchinara

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                baseUrl, userId);
+        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+    }
+
+    protected List<GetUniverseAllocationResponse> handleResponse(Response response) throws Exception {
+        String body = response.body().string();
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:44]: kalau response-nya null gimana? @clchinara

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public GetUniverseAllocationResponse getFallback() {
+        return null;
+    }
+
+    @Override
+    protected Request createRequest() {
+        String url = String.format("%s/v1/abtest/allocation/%s/%s",
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:38]: jadiin constant di class ini @clchinara

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                baseUrl, userId, universeId);
+        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+    }
+
+    protected GetUniverseAllocationResponse handleResponse(Response response) throws Exception {
+        String body = response.body().string();
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:44]: kalau response null gimana?

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        if (!response.isSuccessful()) {
+            throw new KaleException("Failed response from kale status: " +
+                    response.code() + " body: " + response.body().string());
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:47]: kalau response null gimana?

File: src/test/java/com/sayurbox/kale/KaleABTestTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+                .build();
+        ABTestClient abTest = new ABTestClientImpl(config);
+
+        GetUniverseAllocationResponse universeAllocation = abTest.getUniverseAllocation(
+                "qFA88l70U4RxscuJZFUwEZpHUUUF", "0b95519c-1b32-47fe-aa8c-55cb65d6f8c4"
+        );
+        List<GetUniverseAllocationResponse> allUniverseAllocations = abTest.getAllUniverseAllocations(
+                "qFA88l70U4RxscuJZFUwEZpHUUUF"
+        );
+        System.out.println(universeAllocation);
```
Comment (arifpz) [inline src/test/java/com/sayurbox/kale/KaleABTestTest.java:30]: gak pake slf4j annotation aja? instead pakai sysout

File: src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    protected Request createRequest() {
+        String url = String.format(ENDPOINT, baseUrl, userId);
+        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+    }
+
+    protected List<GetUniverseAllocationResponse> handleResponse(Response response) throws Exception {
+        if (response.body() == null) {
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetAllUniverseAllocationsCommand.java:45]: Kalau gini gimana handlingnya:
```
response == null
```

File: src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    protected Request createRequest() {
+        String url = String.format(ENDPOINT, baseUrl, userId, universeId);
+        return new Request.Builder().post(RequestBody.create(null, new byte[]{})).url(url).build();
+    }
+
+    protected GetUniverseAllocationResponse handleResponse(Response response) throws Exception {
+        if (response.body() == null) {
```
Comment (arifpz) [inline src/main/java/com/sayurbox/kale/abtest/command/GetUniverseAllocationCommand.java:45]: Kalau gini apa yg terjadi: response == null

File: build.gradle [hunk 14]
```diff
@@ -31,6 +31,9 @@ dependencies {
     implementation "org.slf4j:slf4j-api:1.7.25"
     implementation "com.squareup.okhttp3:okhttp:3.14.6"
     implementation "com.google.code.gson:gson:2.8.5"
+    implementation 'org.projectlombok:lombok:1.18.20'
```
Comment (andrisasuke) [inline build.gradle:34]: gk perlu pake lombok harusnya.

**Comment by clchinara** [comment]
Reverted null checks on response.body() because it is non-null in this case
(referring to: https://javadoc.io/doc/com.squareup.okhttp3/okhttp/3.14.6/index.html).

File: build.gradle (hunk 14 above)
Comment (clchinara) [reply build.gradle]: Utk Slf4j kakk

Comment (andrisasuke) [reply build.gradle]: private static final Logger logger = LoggerFactory.getLogger(NamaClass.class);
logger.info()
logger.error()

pake yg model gitu aja,  udah ada dari lib `org.slf4j:slf4j-api:`

**Review by andrisasuke — APPROVED** [review]

**Pull request merged**